src/typechecker.erl

```diff
@@ -2987,24 +2982,42 @@ refine_ty(Ty, {Tag, _, V}, TEnv) when ?is_int_type(Ty),
                           int_range_to_types({V + 1, Hi}),
                     normalize(type(union, Tys), TEnv);
                 true ->
-                    Ty
+                    Ty %% disjoint (V is not in the range)
             end;
-        _ ->
+        _OtherIntType ->
             %% Not possible to subtract from other integer types
             throw(no_refinement)
     end;
 refine_ty(Ty1, Ty2, TEnv) ->
-    case subtype(Ty2, Ty1, TEnv) of
-        false     -> Ty1;                 %% disjoint
-        {true, _} -> throw(no_refinement) %% inexact
+    case glb(Ty1, Ty2, TEnv) of
+        ?type(none)  -> Ty1;                 %% disjoint
+        _NotDisjoint -> throw(no_refinement) %% imprecision
     end.
 
-%% [Xs, Ys, ...] -> [[X, Y, ...] || X <- Xs, Y <- Ys, ...]
-pick_from_each([]) -> [];
-pick_from_each([Xs]) ->
-    [[X] || X <- Xs];
-pick_from_each([Xs|Xss]) ->
-    [[X | Ys] || X <- Xs, Ys <- pick_from_each(Xss)].
+%% Returns a nested list on the form
+%%
+%%     [[RefTy1, Ty2, Ty3, ..., TyN],
+%%      [Ty1, RefTy2, Ty3, ..., TyN],
+%%      [Ty1, Ty2, RefTy3, ..., TyN],
+%%      ...
+%%      [Ty1, Ty2, Ty3, ..., RefTyN]].
+%%
+%% If RefTyI==TyI or RefTyI==none() for any I, that list I is excluded.
+pick_one_refinement_each([], []) -> [];
+pick_one_refinement_each([Ty|Tys], [RefTy|RefTys]) ->
+    %% The lists (zero or one list) where we refine head and keep tail
+    %% unrefined, if head is possible to refine.
+    RefHeadCombinations =
+        case RefTy of
+            Ty          -> [];           %% pattern mismatches; no refinement
+            ?type(none) -> [];           %% pattern matches all; no refinement
```

Comment:
I'm still not happy with disjoint pattern and complete match being treated the same. With this implementation you get:
```erlang
refine({a, b | c}, {b, c}) -> {a, b}.
```
since the mismatch of the first component is treated as a complete match. Now `{b, c}` isn't a well-typed pattern of type `{a, b | c}`, but from this you get
```erlang
refine({a, b | c} | {b, c}, {b, c}) -> {a, b}.
```
which is clearly wrong.

Currently I don't think you can get there because of https://github.com/josefs/Gradualizer/blob/9fa405f1dd470c4f0effc269379e8b14c7dee6ef/src/typechecker.erl#L3135
But if that TODO is addressed we would get incorrect refinements.

I think the right thing to do is to skip this case, i.e. fall through to the `[[RefTy|Tys]]` case. That would give us
```erlang
refine({a, b | c}, {b, c}) -> {a, b | c} | {a, b}.
```
which is correct (although the `{a, b}` component looks a little weird).

For products of sets we have A₁ × B₁ - A₂ × B₂ = (A₁ - A₂) × B₁ ∪ A₁ × (B₁ - B₂), or visually:

```
           A₂
        +------+
        |      |
   +----|--+   | B₂ 
   |////|  |   |
B₁ |////+--+---+
   |XXXX\\\|
   +-------+
       A₁
```

In the code the `none()` case is simply an optimisation for when A₁ - A₂ = ∅, and it would be perfectly fine to always return `[[RefTy|Tys]]`.

Reply:
Thanks @UlfNorell for the thorough thoughts! I'll try removing that clause and see if the tests still pass. I'll also add tests for the examples you mention here. I can do this tonight, otherwise in the weekend.

It would be very nice if someone (you?😉) wants to address the `%% TODO: This code approximates unions of tuples with tuples of unions`. It would be a relief to get rid of those.

@josefs I've added a wiki page about [Type refinement](https://github.com/josefs/Gradualizer/wiki/Type-refinement). I didn't put much effort in it, but it's something. Feel free to comment on it here or otherwise or edit it yourself.

Reply:
The wiki page is great!

Reply:
I've been processing the disjoint pattern case in my sleep.

If a pattern inside a (nested) tuple pattern can't match, it means the whole tuple pattern can't match. Therefore, I think the right thing to do is a non-local return to an outer union, where the refinement is applied. If there is no union, the exception can be caught in `type_check_clause` and a message "this clause can never match" can be reported.

Reply:
Disjoint is now thrown upward (caught where a union is refined) so at least that case is gone. Tuple union refinement now behaves nicely, e.g. `refine({a, b | c}, {b, c}) -> {a, b | c}.`

If the whole thing is refined to `none()`, it means the next clause is unreachable, so we have unreachable clause detection implemented as well as a bonus.